### PR TITLE
Flintpoly

### PR DESCRIPTION
**Pull request by srajangarg**
`to_integer_class` methods also added.

### CODE REVIEW

**Comment by srajangarg** [comment]
@isuruf @Sumith1896 PR coming to an end, please have a look

File: symengine/mp_class.h [hunk 1]
```diff
@@ -4,14 +4,18 @@
 #include <symengine/symengine_config.h>
 #include <symengine/mp_wrapper.h>
 
-#if SYMENGINE_INTEGER_CLASS == SYMENGINE_PIRANHA
+#ifdef HAVE_SYMENGINE_PIRANHA
 #include <piranha/mp_integer.hpp>
 #include <piranha/mp_rational.hpp>
```
Comment (isuruf) [inline symengine/mp_class.h]: Having these headers slows down the compilation right?

Comment (srajangarg) [inline symengine/mp_class.h]: Yes, probably. But these are needed for `to_integer_class` conversions to work. Consider the case when `integer_class` is `fmpz_wrapper` and the user is using `UIntPolyPiranha`. We need a mechanism to convert from `piranha::integer` to `fmpz_wrapper`, and thus piranha functions are required.

Same argument for the `#if SYMENGINE_INTEGER_CLASS == SYMENGINE_FLINT` -> `#ifdef HAVE_SYMENGINE_FLINT` change.

Comment (isuruf) [inline symengine/mp_class.h]: What I want to highlight is that this header is included everywhere. So, your changes will have to be moved to another file.

Comment (srajangarg) [inline symengine/mp_class.h]: Where can they be moved? `to_integer_class` is used in `printer.cpp` as well.

Comment (isuruf) [inline symengine/mp_class.h]: Move it to someplace like `polys/uintbase.h`

Comment (isuruf) [inline symengine/mp_class.h]: >Edit : Does this change really affect compilation time that much? 

Why don't you check it out and report back?

**Comment by srajangarg** [comment]
@isuruf the new code is in `upolybase.h`. Please have a look. It's very untidy right now, can be cleaned up. make after `cmake -DBUILD_BENCHMARKS=no -DWITH_PIRANHA=yes -DINTEGER_CLASS=flint .'

File: symengine/polys/upolybase.h [hunk 2]
```diff
@@ -285,6 +286,65 @@ RCP<const Poly> mul_upoly(const Poly &a, const Poly &b)
     dict *= b.get_poly();
     return Poly::from_container(a.get_var(), std::move(dict));
 }
+
+// misc methods
+#if SYMENGINE_INTEGER_CLASS == SYMENGINE_GMPXX || SYMENGINE_INTEGER_CLASS == SYMENGINE_GMP
+#ifdef HAVE_SYMENGINE_FLINT
+#include <flint/flint.h>
```
Comment (isuruf) [inline symengine/polys/upolybase.h]: You are including this header inside `SymEngine` namespace

**Comment by srajangarg** [comment]
Before moving `to_integer_class`
```
cmake -DBUILD_BENCHMARKS=no -DWITH_PIRANHA=yes -DINTEGER_CLASS=flint .
time make
# real	5m45.404s
# user	5m24.272s
# sys	0m19.104s

cmake -DBUILD_BENCHMARKS=no -DWITH_FLINT=yes -DINTEGER_CLASS=piranha .
time make
# real	6m17.241s
# user	6m03.092s
# sys	0m21.184s
```
After,
```
cmake -DBUILD_BENCHMARKS=no -DWITH_PIRANHA=yes -DINTEGER_CLASS=flint .
time make
# real	6m38.020s
# user	6m13.296s
# sys	0m22.540s

cmake -DBUILD_BENCHMARKS=no -DWITH_FLINT=yes -DINTEGER_CLASS=piranha .
time make
# real	6m47.364s
# user	6m22.380s
# sys	0m22.896s
```
Makes a good amount of difference.

**Comment by isuruf** [comment]
For me, compilation time increases from 3m45s to 4m13s with `integer_class=flint`

**Comment by srajangarg** [comment]
After moving to `upolybase.h`? How is that possible?

**Comment by isuruf** [comment]
I meant the other way around

**Comment by srajangarg** [comment]
I didn't get you, you mean if we replace `#if INTEGER_CLASS == *` with `#ifdef HAVE_SYEMENGINE_*` your compilation time increased?

**Comment by isuruf** [comment]
Yes

**Comment by srajangarg** [comment]
@isuruf any other changes required?

File: symengine/polys/upolybase.h [hunk 3]
```diff
@@ -269,8 +301,55 @@ RCP<const Poly> mul_upoly(const Poly &a, const Poly &b)
...
+}
+#endif
+
+#elif SYMENGINE_INTEGER_CLASS == SYMENGINE_PIRANHA
+#ifdef HAVE_SYMENGINE_FLINT
+inline integer_class to_integer_class(const flint::fmpzxx &i)
+{
+    mpz_t x;
+    fmpz_get_mpz(x, i._data().inner);
+    return integer_class(x);
```
Comment (isuruf) [inline symengine/polys/upolybase.h]: Both `integer_class` and `fmpzxx` use `fmpz_t` as storage. Do a direct copy without going through mpz.

File: symengine/polys/upolybase.h [hunk 4]
```diff
@@ -269,8 +301,55 @@ RCP<const Poly> mul_upoly(const Poly &a, const Poly &b)
...
+{
+    return integer_class(x.get_mpz_view());
+}
+#endif
+
+inline integer_class to_integer_class(const flint::fmpzxx &i)
+{
+    mpz_t x;
+    fmpz_get_mpz(x, i._data().inner);
+    return integer_class(x);
```
Comment (isuruf) [inline symengine/polys/upolybase.h]: Both integer_class and fmpzxx use fmpz_t as storage. Do a direct copy without going through mpz.

File: symengine/polys/upolybase.h (hunk 3 above)
Comment (isuruf) [inline symengine/polys/upolybase.h]: Ignore this

Comment (isuruf) [inline symengine/polys/upolybase.h]: `mpz_t` have to be initialized first and then cleared.

Comment (isuruf) [inline symengine/polys/upolybase.h]: Declare `x` as a `integer_class`

File: symengine/printer.cpp [hunk 5]
```diff
@@ -357,6 +357,70 @@ void StrPrinter::bvisit(const UIntPoly &x)
...
+        // if exponent is not 1, print the exponent;
+        if (i != 1) {
+            s << "**" << i;
+        }
+        // corner cases of only first term handled successfully, switch the bool
+        first = false;
+    }
+    if (p.length() == 0)
+        s << "0";
+    str_ = s.str();
```
Comment (isuruf) [inline symengine/printer.cpp]: Advantage of `UIntPolyBase` should be to be able to write methods like these as one function without duplicating the code for both specializations.

Comment (srajangarg) [inline symengine/printer.cpp]: What do you think can be done, their container are all different. I may have to add several methods in `UPolyBase`.

Comment (isuruf) [inline symengine/printer.cpp]: Maybe writing an iterator class for flint_polyxx would work?

Comment (srajangarg) [inline symengine/printer.cpp]: @isuruf will it be ok if I do this after initial implementation of `UIntPiranha` is complete? I think we'll need iterators (or a similar mechanism) for `print` and `diff`. When I'm done with all the classes, I will have a better idea as to what all is required, what methods are common etc.

File: symengine/polys/upolybase.h [hunk 6]
```diff
@@ -269,8 +315,53 @@ RCP<const Poly> mul_upoly(const Poly &a, const Poly &b)
...
+    return x;
+}
+#endif
+
+#elif SYMENGINE_INTEGER_CLASS == SYMENGINE_PIRANHA
+#ifdef HAVE_SYMENGINE_FLINT
+inline integer_class to_integer_class(const flint::fmpzxx &i)
+{
+    integer_class x;
+    fmpz_get_mpz(x, i._data().inner);
```
Comment (isuruf) [inline symengine/polys/upolybase.h]: `fmpz_get_mpz(get_mpz_t(x), i._data().inner);`

File: symengine/polys/upolybase.h [hunk 7]
```diff
@@ -269,8 +315,53 @@ RCP<const Poly> mul_upoly(const Poly &a, const Poly &b)
...
+    fmpz_get_mpz(x.get_mpz_t(), i._data().inner);
+    return x;
+}
+#endif
+
+#ifdef HAVE_SYMENGINE_PIRANHA
+inline integer_class to_integer_class(const piranha::integer &i)
+{
+    integer_class x;
+    mpz_init_set(x.get_mpz_t(), i.get_mpz_view());
```
Comment (isuruf) [inline symengine/polys/upolybase.h]: Use `mpz_set`. `x` is already initialised

**Pull request merged**